File: src/main/java/com/ptaf/ai/quality/BasicGherkinValidator.java
package com.ptaf.ai.quality;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

/** Phase 3: lightweight structural checks on a single feature file (no full Gherkin parser). */
public final class BasicGherkinValidator {

    private static final Pattern FEATURE = Pattern.compile("^\\s*Feature:\\s*\\S+", Pattern.CASE_INSENSITIVE);
    private static final Pattern SCENARIO = Pattern.compile("^\\s*(Scenario|Scenario Outline):\\s*\\S+", Pattern.CASE_INSENSITIVE);

    private BasicGherkinValidator() {
    }

    public static List<String> validate(String relativePath, String content) {
        List<String> issues = new ArrayList<>();
        String[] lines = content.split("\\R");
        boolean hasFeature = false;
        boolean hasScenario = false;
        for (String line : lines) {
            if (FEATURE.matcher(line).find()) {
                hasFeature = true;
            }
            if (SCENARIO.matcher(line).find()) {
                hasScenario = true;
            }
        }
        if (!hasFeature) {
            issues.add(relativePath + ": missing Feature: line");
        }
        if (!hasScenario) {
            issues.add(relativePath + ": missing Scenario / Scenario Outline");
        }
        return issues;
    }
}

File: python/ptaf/ai/quality/basic_gherkin_validator.py
from __future__ import annotations

import re

_FEATURE = re.compile(r"^\s*Feature:\s*\S+", re.IGNORECASE)
_SCENARIO = re.compile(r"^\s*(Scenario|Scenario Outline):\s*\S+", re.IGNORECASE)


def validate(relative_path: str, content: str) -> list[str]:
    issues: list[str] = []
    has_feature = False
    has_scenario = False
    for line in content.splitlines():
        if _FEATURE.search(line):
            has_feature = True
        if _SCENARIO.search(line):
            has_scenario = True
    if not has_feature:
        issues.append(f"{relative_path}: missing Feature: line")
    if not has_scenario:
        issues.append(f"{relative_path}: missing Scenario / Scenario Outline")
    return issues
